general/mem_manager.hpp

```diff
@@ -26,28 +26,64 @@ namespace mfem
 /// Memory types supported by MFEM.
 enum class MemoryType
 {
-   HOST,      ///< Host memory; using new[] and delete[]
-   HOST_32,   ///< Host memory aligned at 32 bytes (not supported yet)
-   HOST_64,   ///< Host memory aligned at 64 bytes (not supported yet)
-   CUDA,      ///< cudaMalloc, cudaFree
-   CUDA_UVM   ///< cudaMallocManaged, cudaFree (not supported yet)
+   HOST,           ///< Host memory; using new[] and delete[]
+   HOST_32,        ///< Host memory; aligned at 32 bytes
+   HOST_64,        ///< Host memory; aligned at 64 bytes
+   HOST_DEBUG,     ///< Host memory; protected while in device mode
+   HOST_UMPIRE,    ///< Host memory; using Umpire
+   HOST_MANAGED,   ///< Host memory; using *MallocManaged, *Free
+   DEVICE,         ///< Device memory; using *Malloc, *Free
+   DEVICE_DEBUG,   ///< Device memory; protected while in host mode
+   DEVICE_UMPIRE,  ///< Device memory; using Umpire
+   DEVICE_MANAGED, ///< Device memory; using D2D copies only
```

Comment:
I'm don't quite understand: why do we need both `HOST_MANAGED` and `DEVICE_MANAGED`? There is really one memory type, `MANAGED` (or `UVM`).

Reply:
I still think it is better to split `MANAGED` in to `CUDA_MANAGED`/`CUDA_UVM` and `HIP_MANAGED` (or whatever is the right term in hip instead of "managed").

Reply:
It might be possible to merge both `HOST_MANAGED` and `DEVICE_MANAGED`.

However, I didn't want to do a special case for this backend: each of them come with one pair of host/device enums.

And again, I don't think from the user perspective, we should be naming memory backends specifically.

Reply:
I looked at the code using `HOST_MANAGED` and `DEVICE_MANAGED` and it seems that unifying the two is essentially a matter of replacing the two strings with `MANAGED`. I think this simplification is worth it.

Reply:
One thing to be careful about is to put `MANAGED` where `HOST_MANAGED` is right now, so it will be correctly identified as host-type by the function `IsHostMemory()`. Then the check (in `MFEM_VERIFY_TYPES`) for device memory should be `!IsHostMemory(d_mt) || d_mt == MemoryType::MANAGED` which can be also defines as a function `IsDeviceMemory()`.

Reply:
You're right, these are the places to be careful with the merged `MANAGED`.

Reply:
`done`